README.md
rust-uritemplate
================

`rust-uritemplate` is a Rust implementation of
[RFC6570  - URI Template](http://tools.ietf.org/html/rfc6570) that can process
URI Templates up to and including ones designated as Level 4 by the
specification. It passes all of the tests in the
[uritemplate-test](https://github.com/uri-templates/uritemplate-test) test
suite.

Basic Usage
-----------
Variable setting can be chained for nice, clean code.

```rust
let uri = UriTemplate::new("/view/{object:1}/{/object,names}{?query*}")
    .set("object", "lakes")
    .set("names", &["Erie", "Superior", "Ontario"])
    .set("query", &[("size", "15"), ("lang", "en")])
    .build();

assert_eq!(uri, "/view/l/lakes/Erie,Superior,Ontario?size=15&lang=en");
```

It is not possible to set a variable to the value "undefined". Instead, simply
delete the variable if you have already set it.

```rust
let mut t = UriTemplate::new("{hello}");
t.set("hello", "Hello World!");
assert_eq!(t.build(), "Hello%20World%21");

t.delete("hello");
assert_eq!(t.build(), "");
```

The `delete` function returns `true` if the variable existed and `false`
otherwise.

Supported Types
---------------
Any type that implements `IntoTemplateVar` can be set as the value of a
`UriTemplate` variable. The following implementations are provided by default
for each type of variable:

- Scalar Values: `String`, `&str`
- Lists: `Vec<String>`, `&[String]`, `&[str]`
- Associative Arrays: `Vec<(String, String)>`, `&[(String, String)]`,
  `&[(&str, &str)]`, `&HashMap<String, String>`

In addition, you can implement `IntoTemplateVar` for your own types. View the
documentation for `IntoTemplateVar` for information on how that works.


CHANGELOG.md
# Changelog

## 0.1.1 (2015-05-28)

- Literal expansion now encodes any characters that are not
  reserved/unreserved/pct-encoded, as outlined in Section 3.1. Literal Expansion
  of the specification.

## 0.1.0 (2015-05-27)

- Initial release


tests/spec_examples.rs
// File autogenerated with /scripts/generate_tests.py

extern crate uritemplate;
use uritemplate::UriTemplate;

// Level 1 Examples
#[test]
fn test_level_1() {
    let mut templates = [
        UriTemplate::new("{var}"),
        UriTemplate::new("{hello}"),
    ];

    for i in 0..templates.len() {
        templates[i].set("hello", "Hello World!");
        templates[i].set("var", "value");
    }

    assert_eq!(templates[0].build(), "value");
    assert_eq!(templates[1].build(), "Hello%20World%21");
}

// Level 2 Examples
#[test]
fn test_level_2() {
    let mut templates = [
        UriTemplate::new("{+var}"),
        UriTemplate::new("{+hello}"),
        UriTemplate::new("{+path}/here"),
        UriTemplate::new("here?ref={+path}"),
    ];

    for i in 0..templates.len() {
        templates[i].set("path", "/foo/bar");
        templates[i].set("hello", "Hello World!");
        templates[i].set("var", "value");
    }

    assert_eq!(templates[0].build(), "value");
    assert_eq!(templates[1].build(), "Hello%20World!");
    assert_eq!(templates[2].build(), "/foo/bar/here");
    assert_eq!(templates[3].build(), "here?ref=/foo/bar");
}

// Level 3 Examples
#[test]
fn test_level_3() {
    let mut templates = [
        UriTemplate::new("map?{x,y}"),
        UriTemplate::new("{x,hello,y}"),
        UriTemplate::new("{+x,hello,y}"),
        UriTemplate::new("{+path,x}/here"),
        UriTemplate::new("{#x,hello,y}"),
        UriTemplate::new("{#path,x}/here"),
        UriTemplate::new("X{.var}"),
        UriTemplate::new("X{.x,y}"),
        UriTemplate::new("{/var}"),
        UriTemplate::new("{/var,x}/here"),
        UriTemplate::new("{;x,y}"),
        UriTemplate::new("{;x,y,empty}"),
        UriTemplate::new("{?x,y}"),
        UriTemplate::new("{?x,y,empty}"),
        UriTemplate::new("?fixed=yes{&x}"),
        UriTemplate::new("{&x,y,empty}"),
    ];

    for i in 0..templates.len() {
        templates[i].set("empty", "");
        templates[i].set("var", "value");
        templates[i].set("y", "768");
        templates[i].set("path", "/foo/bar");
        templates[i].set("hello", "Hello World!");
        templates[i].set("x", "1024");
    }

    assert_eq!(templates[0].build(), "map?1024,768");
    assert_eq!(templates[1].build(), "1024,Hello%20World%21,768");
    assert_eq!(templates[2].build(), "1024,Hello%20World!,768");
    assert_eq!(templates[3].build(), "/foo/bar,1024/here");
    assert_eq!(templates[4].build(), "#1024,Hello%20World!,768");
    assert_eq!(templates[5].build(), "#/foo/bar,1024/here");
    assert_eq!(templates[6].build(), "X.value");
    assert_eq!(templates[7].build(), "X.1024.768");
    assert_eq!(templates[8].build(), "/value");
    assert_eq!(templates[9].build(), "/value/1024/here");
    assert_eq!(templates[10].build(), ";x=1024;y=768");
    assert_eq!(templates[11].build(), ";x=1024;y=768;empty");
    assert_eq!(templates[12].build(), "?x=1024&y=768");
    assert_eq!(templates[13].build(), "?x=1024&y=768&empty=");
    assert_eq!(templates[14].build(), "?fixed=yes&x=1024");
    assert_eq!(templates[15].build(), "&x=1024&y=768&empty=");
}

// Level 4 Examples
#[test]
fn test_level_4() {
    let mut templates = [
        UriTemplate::new("{var:3}"),
        UriTemplate::new("{var:30}"),
        UriTemplate::new("{list}"),
        UriTemplate::new("{list*}"),
        UriTemplate::new("{keys}"),
        UriTemplate::new("{keys*}"),
        UriTemplate::new("{+path:6}/here"),
        UriTemplate::new("{+list}"),
        UriTemplate::new("{+list*}"),
        UriTemplate::new("{+keys}"),
        UriTemplate::new("{+keys*}"),
        UriTemplate::new("{#path:6}/here"),
        UriTemplate::new("{#list}"),
        UriTemplate::new("{#list*}"),
        UriTemplate::new("{#keys}"),
        UriTemplate::new("{#keys*}"),
        UriTemplate::new("X{.var:3}"),
        UriTemplate::new("X{.list}"),
        UriTemplate::new("X{.list*}"),
        UriTemplate::new("X{.keys}"),
        UriTemplate::new("{/var:1,var}"),
        UriTemplate::new("{/list}"),
        UriTemplate::new("{/list*}"),
        UriTemplate::new("{/list*,path:4}"),
        UriTemplate::new("{/keys}"),
        UriTemplate::new("{/keys*}"),
        UriTemplate::new("{;hello:5}"),
        UriTemplate::new("{;list}"),
        UriTemplate::new("{;list*}"),
        UriTemplate::new("{;keys}"),
        UriTemplate::new("{;keys*}"),
        UriTemplate::new("{?var:3}"),
        UriTemplate::new("{?list}"),
        UriTemplate::new("{?list*}"),
        UriTemplate::new("{?keys}"),
        UriTemplate::new("{?keys*}"),
        UriTemplate::new("{&var:3}"),
        UriTemplate::new("{&list}"),
        UriTemplate::new("{&list*}"),
        UriTemplate::new("{&keys}"),
        UriTemplate::new("{&keys*}"),
    ];

    for i in 0..templates.len() {
        templates[i].set("path", "/foo/bar");
        templates[i].set("hello", "Hello World!");
        templates[i].set("var", "value");
        templates[i].set("list", &["red", "green", "blue"] as &[&str]);
        templates[i].set("keys", &[("dot", "."), ("semi", ";"), ("comma", ",")] as &[(&str, &str)]);
    }

    assert_eq!(templates[0].build(), "val");
    assert_eq!(templates[1].build(), "value");
    assert_eq!(templates[2].build(), "red,green,blue");
    assert_eq!(templates[3].build(), "red,green,blue");
    let template_4_answers = vec![
        "comma,%2C,dot,.,semi,%3B",
        "comma,%2C,semi,%3B,dot,.",
        "dot,.,comma,%2C,semi,%3B",
        "dot,.,semi,%3B,comma,%2C",
        "semi,%3B,comma,%2C,dot,.",
        "semi,%3B,dot,.,comma,%2C",
    ];
    assert!(template_4_answers.contains(&templates[4].build().as_ref()));
    let template_5_answers = vec![
        "comma=%2C,dot=.,semi=%3B",
        "comma=%2C,semi=%3B,dot=.",
        "dot=.,comma=%2C,semi=%3B",
        "dot=.,semi=%3B,comma=%2C",
        "semi=%3B,comma=%2C,dot=.",
        "semi=%3B,dot=.,comma=%2C",
    ];
    assert!(template_5_answers.contains(&templates[5].build().as_ref()));
    assert_eq!(templates[6].build(), "/foo/b/here");
    assert_eq!(templates[7].build(), "red,green,blue");
    assert_eq!(templates[8].build(), "red,green,blue");
    let template_9_answers = vec![
        "comma,,,dot,.,semi,;",
        "comma,,,semi,;,dot,.",
        "dot,.,comma,,,semi,;",
        "dot,.,semi,;,comma,,",
        "semi,;,comma,,,dot,.",
        "semi,;,dot,.,comma,,",
    ];
    assert!(template_9_answers.contains(&templates[9].build().as_ref()));
    let template_10_answers = vec![
        "comma=,,dot=.,semi=;",
        "comma=,,semi=;,dot=.",
        "dot=.,comma=,,semi=;",
        "dot=.,semi=;,comma=,",
        "semi=;,comma=,,dot=.",
        "semi=;,dot=.,comma=,",
    ];
    assert!(template_10_answers.contains(&templates[10].build().as_ref()));
    assert_eq!(templates[11].build(), "#/foo/b/here");
    assert_eq!(templates[12].build(), "#red,green,blue");
    assert_eq!(templates[13].build(), "#red,green,blue");
    let template_14_answers = vec![
        "#comma,,,dot,.,semi,;",
        "#comma,,,semi,;,dot,.",
        "#dot,.,comma,,,semi,;",
        "#dot,.,semi,;,comma,,",
        "#semi,;,comma,,,dot,.",
        "#semi,;,dot,.,comma,,",
    ];
    assert!(template_14_answers.contains(&templates[14].build().as_ref()));
    let template_15_answers = vec![
        "#comma=,,dot=.,semi=;",
        "#comma=,,semi=;,dot=.",
        "#dot=.,comma=,,semi=;",
        "#dot=.,semi=;,comma=,",
        "#semi=;,comma=,,dot=.",
        "#semi=;,dot=.,comma=,",
    ];
    assert!(template_15_answers.contains(&templates[15].build().as_ref()));
    assert_eq!(templates[16].build(), "X.val");
    assert_eq!(templates[17].build(), "X.red,green,blue");
    assert_eq!(templates[18].build(), "X.red.green.blue");
    let template_19_answers = vec![
        "X.comma,%2C,dot,.,semi,%3B",
        "X.comma,%2C,semi,%3B,dot,.",
        "X.dot,.,comma,%2C,semi,%3B",
        "X.dot,.,semi,%3B,comma,%2C",
        "X.semi,%3B,comma,%2C,dot,.",
        "X.semi,%3B,dot,.,comma,%2C",
    ];
    assert!(template_19_answers.contains(&templates[19].build().as_ref()));
    assert_eq!(templates[20].build(), "/v/value");
    assert_eq!(templates[21].build(), "/red,green,blue");
    assert_eq!(templates[22].build(), "/red/green/blue");
    assert_eq!(templates[23].build(), "/red/green/blue/%2Ffoo");
    let template_24_answers = vec![
        "/comma,%2C,dot,.,semi,%3B",
        "/comma,%2C,semi,%3B,dot,.",
        "/dot,.,comma,%2C,semi,%3B",
        "/dot,.,semi,%3B,comma,%2C",
        "/semi,%3B,comma,%2C,dot,.",
        "/semi,%3B,dot,.,comma,%2C",
    ];
    assert!(template_24_answers.contains(&templates[24].build().as_ref()));
    let template_25_answers = vec![
        "/comma=%2C/dot=./semi=%3B",
        "/comma=%2C/semi=%3B/dot=.",
        "/dot=./comma=%2C/semi=%3B",
        "/dot=./semi=%3B/comma=%2C",
        "/semi=%3B/comma=%2C/dot=.",
        "/semi=%3B/dot=./comma=%2C",
    ];
    assert!(template_25_answers.contains(&templates[25].build().as_ref()));
    assert_eq!(templates[26].build(), ";hello=Hello");
    assert_eq!(templates[27].build(), ";list=red,green,blue");
    assert_eq!(templates[28].build(), ";list=red;list=green;list=blue");
    let template_29_answers = vec![
        ";keys=comma,%2C,dot,.,semi,%3B",
        ";keys=comma,%2C,semi,%3B,dot,.",
        ";keys=dot,.,comma,%2C,semi,%3B",
        ";keys=dot,.,semi,%3B,comma,%2C",
        ";keys=semi,%3B,comma,%2C,dot,.",
        ";keys=semi,%3B,dot,.,comma,%2C",
    ];
    assert!(template_29_answers.contains(&templates[29].build().as_ref()));
    let template_30_answers = vec![
        ";comma=%2C;dot=.;semi=%3B",
        ";comma=%2C;semi=%3B;dot=.",
        ";dot=.;comma=%2C;semi=%3B",
        ";dot=.;semi=%3B;comma=%2C",
        ";semi=%3B;comma=%2C;dot=.",
        ";semi=%3B;dot=.;comma=%2C",
    ];
    assert!(template_30_answers.contains(&templates[30].build().as_ref()));
    assert_eq!(templates[31].build(), "?var=val");
    assert_eq!(templates[32].build(), "?list=red,green,blue");
    assert_eq!(templates[33].build(), "?list=red&list=green&list=blue");
    let template_34_answers = vec![
        "?keys=comma,%2C,dot,.,semi,%3B",
        "?keys=comma,%2C,semi,%3B,dot,.",
        "?keys=dot,.,comma,%2C,semi,%3B",
        "?keys=dot,.,semi,%3B,comma,%2C",
        "?keys=semi,%3B,comma,%2C,dot,.",
        "?keys=semi,%3B,dot,.,comma,%2C",
    ];
    assert!(template_34_answers.contains(&templates[34].build().as_ref()));
    let template_35_answers = vec![
        "?comma=%2C&dot=.&semi=%3B",
        "?comma=%2C&semi=%3B&dot=.",
        "?dot=.&comma=%2C&semi=%3B",
        "?dot=.&semi=%3B&comma=%2C",
        "?semi=%3B&comma=%2C&dot=.",
        "?semi=%3B&dot=.&comma=%2C",
    ];
    assert!(template_35_answers.contains(&templates[35].build().as_ref()));
    assert_eq!(templates[36].build(), "&var=val");
    assert_eq!(templates[37].build(), "&list=red,green,blue");
    assert_eq!(templates[38].build(), "&list=red&list=green&list=blue");
    let template_39_answers = vec![
        "&keys=comma,%2C,dot,.,semi,%3B",
        "&keys=comma,%2C,semi,%3B,dot,.",
        "&keys=dot,.,comma,%2C,semi,%3B",
        "&keys=dot,.,semi,%3B,comma,%2C",
        "&keys=semi,%3B,comma,%2C,dot,.",
        "&keys=semi,%3B,dot,.,comma,%2C",
    ];
    assert!(template_39_answers.contains(&templates[39].build().as_ref()));
    let template_40_answers = vec![
        "&comma=%2C&dot=.&semi=%3B",
        "&comma=%2C&semi=%3B&dot=.",
        "&dot=.&comma=%2C&semi=%3B",
        "&dot=.&semi=%3B&comma=%2C",
        "&semi=%3B&comma=%2C&dot=.",
        "&semi=%3B&dot=.&comma=%2C",
    ];
    assert!(template_40_answers.contains(&templates[40].build().as_ref()));
}


tests/test_uritemplate.rs
extern crate uritemplate;
use uritemplate::{IntoTemplateVar, TemplateVar, UriTemplate};

#[test]
fn test_example() {
    let uri = UriTemplate::new("/view/{object:1}{/object,names}{?query*}")
        .set("object", "lakes")
        .set("names", &["Erie", "Superior", "Ontario"])
        .set("query", &[("size", "15"), ("lang", "en")])
        .build();
    assert_eq!(
        uri,
        "/view/l/lakes/Erie,Superior,Ontario?size=15&lang=en"
    );
}

#[test]
fn test_deletion_example() {
    let mut t = UriTemplate::new("{hello}");
    t.set("hello", "Hello World!");
    assert_eq!(t.build(), "Hello%20World%21");

    t.delete("hello");
    assert_eq!(t.build(), "");
}

#[test]
fn test_delete() {
    let mut t = UriTemplate::new("{foo}");
    t.set("foo", "123");

    assert_eq!(t.build(), "123");

    // Deleting nonexistent variable has no effect
    assert_eq!(t.delete("bar"), false);
    assert_eq!(t.build(), "123");

    // Deleting existing variable works
    assert_eq!(t.delete("foo"), true);
    assert_eq!(t.build(), "");
}

#[test]
fn test_delete_all() {
    let mut t = UriTemplate::new("{A}{B}{C}");
    t.set("A", "alpha");
    t.set("B", "beta");
    t.set("C", "charlie");
    assert_eq!(t.build(), "alphabetacharlie");

    t.delete_all();
    assert_eq!(t.build(), "");
}

struct Address {
    city: String,
    state: String
}

impl <'a> IntoTemplateVar for &'a Address {
    fn into_template_var(self) -> TemplateVar {
        TemplateVar::AssociativeArray(vec![
            ("city".to_string(), self.city.clone()),
            ("state".to_string(), self.state.clone())
        ])
    }
}

#[test]
fn test_intotemplatevar() {
    let address = Address {
        city: "Los Angelos".to_string(),
        state: "California".to_string()
    };
    let uri = UriTemplate::new("http://example.com/view{?address*}")
        .set("address", &address)
        .build();
    assert_eq!(
        uri,
        "http://example.com/view?city=Los%20Angelos&state=California"
    );
}

#[test]
fn test_set() {
    let mut t = UriTemplate::new("");
    t.set("hello", "hello".to_string());
    t.set("listvar", Vec::<String>::new());
    t.set("assocvar", Vec::<(String, String)>::new());
}

#[test]
fn test_literal_expansion() {
    let uri = UriTemplate::new("hey!%%25").build();
    assert_eq!(uri, "hey!%25%25");
}


src/percent_encoding.rs
// This module provides support for URL percent encoding as defined in the
// RFC 6570 specification.

use regex::Regex;

static UNRESERVED: [&'static str; 256] = [
   "%00", "%01", "%02", "%03", "%04", "%05", "%06", "%07",
   "%08", "%09", "%0A", "%0B", "%0C", "%0D", "%0E", "%0F",
   "%10", "%11", "%12", "%13", "%14", "%15", "%16", "%17",
   "%18", "%19", "%1A", "%1B", "%1C", "%1D", "%1E", "%1F",
   "%20", "%21", "%22", "%23", "%24", "%25", "%26", "%27",
   "%28", "%29", "%2A", "%2B", "%2C", "-", ".", "%2F",
   "0", "1", "2", "3", "4", "5", "6", "7",
   "8", "9", "%3A", "%3B", "%3C", "%3D", "%3E", "%3F",
   "%40", "A", "B", "C", "D", "E", "F", "G",
   "H", "I", "J", "K", "L", "M", "N", "O",
   "P", "Q", "R", "S", "T", "U", "V", "W",
   "X", "Y", "Z", "%5B", "%5C", "%5D", "%5E", "_",
   "%60", "a", "b", "c", "d", "e", "f", "g",
   "h", "i", "j", "k", "l", "m", "n", "o",
   "p", "q", "r", "s", "t", "u", "v", "w",
   "x", "y", "z", "%7B", "%7C", "%7D", "~", "%7F",
   "%80", "%81", "%82", "%83", "%84", "%85", "%86", "%87",
   "%88", "%89", "%8A", "%8B", "%8C", "%8D", "%8E", "%8F",
   "%90", "%91", "%92", "%93", "%94", "%95", "%96", "%97",
   "%98", "%99", "%9A", "%9B", "%9C", "%9D", "%9E", "%9F",
   "%A0", "%A1", "%A2", "%A3", "%A4", "%A5", "%A6", "%A7",
   "%A8", "%A9", "%AA", "%AB", "%AC", "%AD", "%AE", "%AF",
   "%B0", "%B1", "%B2", "%B3", "%B4", "%B5", "%B6", "%B7",
   "%B8", "%B9", "%BA", "%BB", "%BC", "%BD", "%BE", "%BF",
   "%C0", "%C1", "%C2", "%C3", "%C4", "%C5", "%C6", "%C7",
   "%C8", "%C9", "%CA", "%CB", "%CC", "%CD", "%CE", "%CF",
   "%D0", "%D1", "%D2", "%D3", "%D4", "%D5", "%D6", "%D7",
   "%D8", "%D9", "%DA", "%DB", "%DC", "%DD", "%DE", "%DF",
   "%E0", "%E1", "%E2", "%E3", "%E4", "%E5", "%E6", "%E7",
   "%E8", "%E9", "%EA", "%EB", "%EC", "%ED", "%EE", "%EF",
   "%F0", "%F1", "%F2", "%F3", "%F4", "%F5", "%F6", "%F7",
   "%F8", "%F9", "%FA", "%FB", "%FC", "%FD", "%FE", "%FF",
];

static RESERVED: [&'static str; 256] = [
   "%00", "%01", "%02", "%03", "%04", "%05", "%06", "%07",
   "%08", "%09", "%0A", "%0B", "%0C", "%0D", "%0E", "%0F",
   "%10", "%11", "%12", "%13", "%14", "%15", "%16", "%17",
   "%18", "%19", "%1A", "%1B", "%1C", "%1D", "%1E", "%1F",
   "%20", "!", "%22", "#", "$", "%25", "&", "'",
   "(", ")", "*", "+", ",", "-", ".", "/",
   "0", "1", "2", "3", "4", "5", "6", "7",
   "8", "9", ":", ";", "%3C", "=", "%3E", "?",
   "@", "A", "B", "C", "D", "E", "F", "G",
   "H", "I", "J", "K", "L", "M", "N", "O",
   "P", "Q", "R", "S", "T", "U", "V", "W",
   "X", "Y", "Z", "[", "%5C", "]", "%5E", "_",
   "%60", "a", "b", "c", "d", "e", "f", "g",
   "h", "i", "j", "k", "l", "m", "n", "o",
   "p", "q", "r", "s", "t", "u", "v", "w",
   "x", "y", "z", "%7B", "%7C", "%7D", "~", "%7F",
   "%80", "%81", "%82", "%83", "%84", "%85", "%86", "%87",
   "%88", "%89", "%8A", "%8B", "%8C", "%8D", "%8E", "%8F",
   "%90", "%91", "%92", "%93", "%94", "%95", "%96", "%97",
   "%98", "%99", "%9A", "%9B", "%9C", "%9D", "%9E", "%9F",
   "%A0", "%A1", "%A2", "%A3", "%A4", "%A5", "%A6", "%A7",
   "%A8", "%A9", "%AA", "%AB", "%AC", "%AD", "%AE", "%AF",
   "%B0", "%B1", "%B2", "%B3", "%B4", "%B5", "%B6", "%B7",
   "%B8", "%B9", "%BA", "%BB", "%BC", "%BD", "%BE", "%BF",
   "%C0", "%C1", "%C2", "%C3", "%C4", "%C5", "%C6", "%C7",
   "%C8", "%C9", "%CA", "%CB", "%CC", "%CD", "%CE", "%CF",
   "%D0", "%D1", "%D2", "%D3", "%D4", "%D5", "%D6", "%D7",
   "%D8", "%D9", "%DA", "%DB", "%DC", "%DD", "%DE", "%DF",
   "%E0", "%E1", "%E2", "%E3", "%E4", "%E5", "%E6", "%E7",
   "%E8", "%E9", "%EA", "%EB", "%EC", "%ED", "%EE", "%EF",
   "%F0", "%F1", "%F2", "%F3", "%F4", "%F5", "%F6", "%F7",
   "%F8", "%F9", "%FA", "%FB", "%FC", "%FD", "%FE", "%FF",
];

pub fn encode_unreserved(s: &str) -> String {
    let mut res = String::new();
    for &byte in s.as_bytes() {
        res.push_str(UNRESERVED[byte as usize]);
    }
    res
}

pub fn encode_reserved(s: &str) -> String {
    let mut res = String::new();
    for &byte in s.as_bytes() {
        res.push_str(RESERVED[byte as usize]);
    }
    // Correct any percent-encoded triplets whose percent signs were reencoded
    Regex::new("%25(?P<hex>[0-9a-fA-F][0-9a-fA-F])")
        .unwrap()
        .replace_all(&res, "%$hex")
}


src/templatevar.rs
use std::collections::HashMap;

/// TemplateVar represents the value of a template variable, which can be a
/// scalar (a simple string), a list of strings, or an associative array of
/// strings.
///
/// Normally, it is not necessary to use this class unless you are implementing
/// the `IntoTemplateVar` trait for your own classes.
#[derive(Clone)]
pub enum TemplateVar {
    /// A simple string such as `"foo"`
    Scalar(String),
    /// A list of strings such as `["foo", "bar"]`
    List(Vec<String>),
    /// An associative array of strings, such as
    /// `[("key1", "value1"), ("key2", "value2")]`
    AssociativeArray(Vec<(String, String)>)
}

/// IntoTemplateVar represents any type that can be converted into a TemplateVar
/// for use as the value of a template variable, such as a `String`,
/// `Vec<String>`, or `Vec<(String, String)>`. Default implementations are
/// available for those three types, as well as the `&str` versions.
///
/// Example
/// -------
/// Here is an example of implementing `IntoTemplateVar` for your own `Address`
/// struct. Note that you should probably implement the trait for a reference to
/// your struct, not the actual struct, unless you intend to move the value
/// efficiently.
///
/// ```ignore
/// struct Address {
///     city: String,
///     state: String
/// }
///
/// impl <'a> IntoTemplateVar for &'a Address {
///     fn into_template_var(self) -> TemplateVar {
///         TemplateVar::AssociativeArray(vec![
///             ("city".to_string(), self.city.clone()),
///             ("state".to_string(), self.state.clone())
///         ])
///     }
/// }
/// ```
///
/// Now, `Address` variables can be set as `UriTemplate` variables.
///
/// ```ignore
/// let address = Address {
///     city: "Los Angelos".to_string(),
///     state: "California".to_string()
/// };
///
/// let uri = UriTemplate::new("http://example.com/view{?address*}")
///     .set("address", &address)
///     .build();
/// 
/// assert_eq!(
///     uri,
///     "http://example.com/view?city=Los%20Angelos&state=California"
/// );
/// ```
pub trait IntoTemplateVar {
    fn into_template_var(self) -> TemplateVar;
}

impl IntoTemplateVar for TemplateVar {
    fn into_template_var(self) -> TemplateVar {
        self.clone()
    }
}

impl <'a> IntoTemplateVar for &'a str {
    fn into_template_var(self) -> TemplateVar {
        TemplateVar::Scalar(self.to_string())
    }
}

impl IntoTemplateVar for String {
    fn into_template_var(self) -> TemplateVar {
        TemplateVar::Scalar(self)
    }
}

impl <'a> IntoTemplateVar for &'a [String] {
    fn into_template_var(self) -> TemplateVar {
        let mut vec = Vec::new();
        for s in self {
            vec.push(s.clone());
        }
        TemplateVar::List(vec)
    }
}

impl IntoTemplateVar for Vec<String> {
    fn into_template_var(self) -> TemplateVar {
        TemplateVar::List(self)
    }
}

impl <'a, 'b> IntoTemplateVar for &'a[&'b str] {
    fn into_template_var(self) -> TemplateVar {
        let mut vec = Vec::new();
        for s in self {
            vec.push(s.to_string());
        }
        TemplateVar::List(vec)
    }
}

impl <'a> IntoTemplateVar for &'a [(String, String)] {
    fn into_template_var(self) -> TemplateVar {
        let mut vec = Vec::new();
        for s in self {
            vec.push(s.clone());
        }
        TemplateVar::AssociativeArray(vec)
    }
}

impl IntoTemplateVar for Vec<(String, String)> {
    fn into_template_var(self) -> TemplateVar {
        TemplateVar::AssociativeArray(self)
    }
}

impl <'a, 'b, 'c> IntoTemplateVar for &'a[(&'b str, &'c str)] {
    fn into_template_var(self) -> TemplateVar {
        let mut vec = Vec::new();
        for s in self {
            vec.push((s.0.to_string(), s.1.to_string()));
        }
        TemplateVar::AssociativeArray(vec)
    }
}

impl <'a> IntoTemplateVar for &'a HashMap<String, String> {
    fn into_template_var(self) -> TemplateVar {
        let mut vec = Vec::new();
        for (k, v) in self {
            vec.push((k.clone(), v.clone()));
        }
        TemplateVar::AssociativeArray(vec)
    }
}

impl <'a, 'b, 'c> IntoTemplateVar for &'a HashMap<&'b str, &'c str> {
    fn into_template_var(self) -> TemplateVar {
        let mut vec = Vec::new();
        for (k, v) in self {
            vec.push((k.to_string(), v.to_string()));
        }
        TemplateVar::AssociativeArray(vec)
    }
}

macro_rules! array_impls {
    ($($N:expr)+) => {$(
        impl <'a> IntoTemplateVar for &'a [String; $N] {
            fn into_template_var(self) -> TemplateVar {
                self[..].into_template_var()
            }
        }

        impl <'a, 'b> IntoTemplateVar for &'a[&'b str; $N] {
            fn into_template_var(self) -> TemplateVar {
                self[..].into_template_var()
            }
        }

        impl <'a> IntoTemplateVar for &'a [(String, String); $N] {
            fn into_template_var(self) -> TemplateVar {
                self[..].into_template_var()
            }
        }

        impl <'a, 'b, 'c> IntoTemplateVar for &'a[(&'b str, &'c str); $N] {
            fn into_template_var(self) -> TemplateVar {
                self[..].into_template_var()
            }
        }
    )+}
}

array_impls!(0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
    26 27 28 29 30 31 32);
